Draft: Workspace Provider Docs (#149)
cdr/docs: admin/environment-management/workspace-providers.md

state: beta
---

> Workspace Providers are logical groups of resources developers can deploy

Reviewer: Pardon me for, I'm just thinking out loud here... I suggest rephrasing as one of the following:

* Workspace Providers are logical groups of resources into which developers can deploy workspaces
* Workspace Providers are logical groups of resources that contain developers' workspaces.

I am also not sure if we should be using "environments" in place of "workspaces" here?

This avoids [ending the sentence with a preposition](https://www.grammarly.com/blog/youve-been-lied-to-heres-why-you-absolutely-can-end-a-sentence-with-a-preposition/) (I am also not an English professor and had to look this up)

I would defer to our Word Nerd @khorne3, of course :nerd_face:

Reviewer 2: I'd probably go with

> Workspace Providers are logical groups of resources *to* which developers can deploy workspaces

workspaces onto. Workspace Providers enable a single Coder deployment to
provision and manage workspaces across multiple Kubernetes clusters and

> namespaces, including ones located in other geographies, regions, or clouds.

Reviewer: ones -> those that are?

Distributed teams can use this feature to allow users to manage workspaces in
the cluster that's nearest, reducing network latency and improving developers'
experience. Additionally, you can use workspace providers to support data
sovereignty requirements or increase the isolation between workspaces running in
the same region or cluster.

### Built-In Workspace Provider

> By default, all Coder deployments will have a `built-in` Workspace Provider that

Reviewer: I love the rename from Local -> Built-In, very thoughtful :heart:

> identifies the kubernetes cluster the Coder deployment is in. This enables

Reviewer: ```suggestion
specifies the Kubernetes cluster that contains the Coder deployment. This enables
```

Just a general thought, but perhaps we could think of "Coder Deployment" as a logical thing encompassing both our control plane (CEManager, etc) as well as our data plane (all running Workspaces/Environments). So what we're really saying is that the built-in Workspace Provider identifies the Kubernetes cluster that is hosting the Coder control plane itself, whereas the workspaces and workspace providers are part of our "data plane"?

The distinction is used in telecom to differentiate between elements that are in the critical path (e.g. switches that handle every packet, aka data plane, where performance is absolutely critical) versus elements that are outside the critical path (e.g. management of the data plane layer aka control plane, which is outside the critical path and can therefore be slower/more compute intensive).

[Cloudflare explains it with this analogy](https://www.cloudflare.com/learning/network-layer/what-is-the-control-plane/):

>Think of the control plane as being like the stoplights that operate at the intersections of a city. Meanwhile, the data plane (or the forwarding plane) is more like the cars that drive on the roads, stop at the intersections, and obey the stoplights.

> workspaces to be deployed into the same cluster as the Coder deployment with no

Reviewer: Suggest using active voice, "This allows users to create workspaces in the same cluster as the Coder deployment with no additional configuration."

additional actions required.

### Remote Workspace Providers

> A Workspace Provider can be deployed into any existing kubernetes cluster, and

Reviewer: Question: does "Workspace Provider" refer to the thing providing the infrastructure (e.g. a Kubernetes cluster) or to the agent connected to the thing providing the infrastructure? This reads like the latter, but from a users' perspective, Kubernetes or EC2 is providing the workspaces, right?

Perhaps:

> You can deploy a Workspace Provider in any existing Kubernetes cluster, and developers will be able to create workspaces in that infrastructure, provided that they have permissions to do so.

Author: Sounds like this docs could use an architecture diagram. In my mind, "Workspace Provider" is the functional unit, made up of a envproxy and a cluster we deploy into. I just am hesitant to add in terminology for the envproxy itself, so I wanted to keep it abstract and say "deploy the thing and things work"

will enable that cluster to become a selectable pool of resources developers can
create workspaces in. Remote Workspace Providers can provide lower latency to
developers by locating their workspaces closer to them geographically, or can be
used for workload isolation purposes. See
[Deploying A Workspace Provider](../../guides/deploying-workspace-provider.md)

> to learn how you can expand your Coder deployment to additional kubernetes

Reviewer: kubernetes -> Kubernetes
